DDL statements in query history (#470)

DanCodedThis opened the pull request:
Closes #470

- Schemas created using the query engine
- Databases cannot be created through the query engine yet
- No `public` or `information` schema creation for databases for now

osipovartem commented on crates/runtime/src/http/ui/schemas/handlers.rs:
```diff
@@ -64,25 +65,61 @@ pub struct ApiDoc;
 )]
 #[tracing::instrument(level = "debug", skip(state), err, ret(level = tracing::Level::TRACE))]
 pub async fn create_schema(
+    DFSessionId(session_id): DFSessionId,
```
Why do we need such thing? create schema is already a part of query engine and history already creating for it
why we cannot just send "create schema" in sql editor?
and remove create schema endpoint from UI:)

DanCodedThis replied:
It's a button on the UI, but I agree, we can send a request from UI, to the query endpoint no need for this, then all the "heavy" lifting will be done by UI

@nikitastryuk what do you think?

rampage644 replied:
@osipovartem We do not want to send any SQL queries from the frontend and effectively keep ui-specific API separate from other API (like `/v1/query` etc). 

This way all the logic to interact with the engine stays on the backend part in a single place.

osipovartem replied:
Create schema is not UI specific query. Since we have to support it for snowflake dialect we will have two places where we can create a schema. What about other catalogs?  If we going to add external catalogs, we have to write a lot of logic to support schema creation via UI for s3 tables, and other.

osipovartem replied:
I don't mean send query from ui, I mean that user should open sql-editor and write query, as usual, to create table, schema, external snowflake storage, etc.

rampage644 replied:
My point was "Send `CREATE SCHEMA` statement from the backend not the frontend".

DanCodedThis replied:
> I don't mean send query from ui, I mean that user should open sql-editor and write query, as usual, to create table, schema, external snowflake storage, etc.

As far as I know this endpoint is used in 1-2 places where the `schema` (also `DB`s, but not `table`s) are created through a form (like in Snowflake UI)

DanCodedThis replied:
> Create schema is not UI specific query. Since we have to support it for snowflake dialect we will have two places where we can create a schema. What about other catalogs? If we going to add external catalogs, we have to write a lot of logic to support schema creation via UI for s3 tables, and other.

This is a valid point to velocity

> > I don't mean send query from ui, I mean that user should open sql-editor and write query, as usual, to create table, schema, external snowflake storage, etc.
> 
> As far as I know this endpoint is used in 1-2 places where the `schema` (also `DB`s, but not `table`s) are created through a form (like in Snowflake UI)

We can remove those, but should we though? 

cc @rampage644

rampage644 replied:
> I don't mean send query from ui, I mean that user should open sql-editor and write query, as usual, to create table, schema, external snowflake storage, etc.

This is possible, the original requirement comes from making a UI as accessible to non-tech savvy users as it is in Snowflake (and thus it has different ways to create a schema, via sql and not). 

This original requirement still holds.

rampage644 commented on crates/runtime/src/http/ui/schemas/handlers.rs:
```diff
@@ -64,25 +65,61 @@ pub struct ApiDoc;
+        payload.name.clone()
+    );
+    //TODO: figure out how to unify this, with the delete, instead of copying code (possibly a generic query history `fn`)
+    let mut query_record = QueryRecord::query_start(&sql_string, None);
```
Is this here just to make sure we record all sql queries? If so - I agree with you previous comment: this should not even be here. 

Instead this handler should know nothing about query history. Since `state.execution_svc` has type `Arc<ExecutionService>` we can wire it during execution to be an instance of `RecordableExecutionService` that wraps `ExecutionService`.

DanCodedThis replied:
Yes ideally this should be a trait

DanCodedThis replied:
> Is this here just to make sure we record all sql queries? If so - I agree with you previous comment: this should not even be here.
> 
> Instead this handler should know nothing about query history. Since `state.execution_svc` has type `Arc<ExecutionService>` we can wire it during http server construction to be an instance of `RecordableExecutionService` that wraps `ExecutionService` and performs actual recording.

Yes, I copied it from our `query` endpoint

YaroslavLitvinov replied:
We were discussing this earlier, how query interface should be looking
https://github.com/Embucket/embucket/issues/425

rampage644 replied:
Okay, happy to approve current approach with the promise this will eventually be replaced. I do agree we do want a flexibility on which query to record, just haven't given a proper thought yet what is the correct interface should be (perhaps moving this logic into recorder makes sense, perhaps not).  

+1 on moving `history` module closer to execution.

DanCodedThis commented:
More context on what was done:
- `Service` is a trait implemented by `struct History` and `struct Execution`
- To avoid circular dependency it is done inside the `runtime` crate
- `QueryContext` now contains the `worksheet_id`, again UI is a dripping roof, that drips down to internals (not good)
- If you have other solutions to pass down `worksheet_id` feel free to suggest it
- Fixed some tests
- Fixed `makeg_app`, `run_binary`, etc
- DDL statements for `Schemas` included here (query history)
- Fixed anything else to adhere to new changes

P.S. Sorry for breaking changes to anybody in advance

rampage644 commented on crates/runtime/src/execution/history.rs:
```diff
@@ -0,0 +1,95 @@
+use embucket_metastore::TableIdent as MetastoreTableIdent;
+use std::sync::Arc;
+
+pub struct History {
```
Rename to better expressiveness (RecordingExecutionService, TrackingExecutionService, InstrumentedExecutionService, LoggingExecutionService)

DanCodedThis replied:
Renamed to RecordingExecutionService

rampage644 commented on crates/runtime/src/execution/service.rs:
```diff
@@ -23,23 +23,47 @@ use uuid::Uuid;
 
-pub struct ExecutionService {
+#[async_trait::async_trait]
+pub trait Service: Send + Sync {
```
CoreExecutionService maybe

DanCodedThis replied:
Renamed

DanCodedThis commented:
Requested changes, cc: @rampage644

DanCodedThis commented:
> > If you have other solutions to pass down worksheet_id feel free to suggest it
> 
> What if we somehow pass either a closure or a builder-something-pattern such that when `.query(sql, execution_context)` it has access to variable from the outer scope (worksheetId in our case)?

It would either break the signature or it will be similar to `ScanIterator` but way way worse (since it would mean that the **_trait_** `fn query` would need to return some struct, and not record batches)

DanCodedThis commented:
> > > If you have other solutions to pass down worksheet_id feel free to suggest it
> > 
> > 
> > What if we somehow pass either a closure or a builder-something-pattern such that when `.query(sql, execution_context)` it has access to variable from the outer scope (worksheetId in our case)?
> 
> It would either break the signature or it will be similar to `ScanIterator` but way way worse (since it would mean that the **_trait_** `fn query` would need to return some struct, and not record batches)

Also, a thing to keep in mind is that stable Rust doesn't allow default values in signatures (but nightly should :) )

DanCodedThis commented:
Maybe we could do something through the `Query object`, but again we either break the signature or the result of the `fn query`.

We could surley, if the `QueryContext` values are `Option`s, make it option it self, meaning even for normal querying we would need to use like a builder pattern, but all our queries require the inside of the `QueryContext` I think, so maybe a bad lead

DanCodedThis commented:
> Maybe we could do something through the `Query object`, but again we either break the signature or the result of the `fn query`.
> 
> We could surley, if the `QueryContext` values are `Option`s, make it option it self, meaning even for normal querying we would need to use like a builder pattern, but all our queries require the inside of the `QueryContext` I think, so maybe a bad lead

I would merge for now, if there are no other suggestion/problems. And later down the line, maybe I or anybody else comes up with something better (like `ScanIterator`)

rampage644 commented on crates/runtime/src/execution/recording_service.rs:
```diff
@@ -0,0 +1,95 @@
+use embucket_metastore::TableIdent as MetastoreTableIdent;
+use std::sync::Arc;
+
+pub struct RecordingExecutionService {
```
Would be great to cover this with tests:

1. happy case, query executed with success
2. happy case, query failed
3. post-processing, failure to store the results

rampage644 commented on crates/runtime/src/execution/recording_service.rs:
```diff
@@ -0,0 +1,95 @@
+use std::sync::Arc;
+
+pub struct RecordingExecutionService {
+    pub execution: Arc<CoreExecutionService>,
```
Can we depend here on `dyn ExecutionService` ?

rampage644 commented:
@DanCodedThis Overall approach looks fine with me. However, this PR is already big enough and is not complete. 

For this PR to land, let's please split it into several smaller ones:
 1. One that groups DDL related changes, metastore -> SQL queries in handlers (#470)
 2. Introduction ExecutionService trait and refactoring `struct ExecutionService` -> `struct CoreExecutionService`
 3. Introduction `RecordingExecutionService` and refactoring handlers to remove unrelated logic from there
 
 Approving this PR only as 1st. Assuming two more are coming as follow ups.

rampage644 reviewed (approved):
approve not to block, assuming split this one into 3 different

DanCodedThis commented:
Moved `ExecutionService` trait to another branch
